backend/python/tapa/floorplan.py

```diff
@@ -0,0 +1,248 @@
+import itertools
+import logging
+import os
+import sys
+from concurrent import futures
+from typing import Optional, Dict, Any, Callable, List, Union, Set, Tuple
+
+import tapa.autobridge
+from haoda.report.xilinx import rtl as report
+from tapa import util
+from tapa.task import Task
+from tapa.verilog import xilinx as rtl
+
+from .instance import Instance
+
+_logger = logging.getLogger().getChild(__name__)
+
+class InputError(Exception):
+  pass
+
+
+def get_cpp(cpp_dir: str, name: str) -> str:
+  return os.path.join(cpp_dir, name + '.cpp')
+
+
+def get_floorplan(
+    directive: Optional[Dict[str, Any]],
+    enable_synth_util: bool,
+    work_dir: str,
+    cpp_dir: str,
+    rtl_dir: str,
+    top_task: Task,
+    ctrl_instance_name: str,
+    task_getter: Callable[[str], Task],
+    fifo_width_getter: Callable[[Task, str], int],
+    **kwargs,
+) -> Tuple[Dict[str, str], Dict[str, int], List[str]]:
+  """
+  get the target region of each vertex
+  get the pipeline level of each edge
+  TODO: remove the "directive" input, which is confusing
+  Repalce it by the "floorplan_pre_assignments" option
+  """
+  # run logic synthesis to get an accurate area estimation
+  if enable_synth_util:
+    get_post_synth_area(
+      work_dir,
+      cpp_dir,
+      rtl_dir,
+      directive,
+      top_task,
+      task_getter,
+    )
+
+  _logger.info('generating partitioning constraints')
+  v2s, s2v, topology = tapa.autobridge.generate_floorplan(
+      top_task,
+      work_dir,
+      fifo_width_getter,
+      directive['connectivity'],
+      directive['part_num'],
+      **kwargs,
+  )
+
+  # extract the mapping from vertice to region
+  # inst_to_region = _get_instance_to_region(floorplan)
+  check_floorplan_validity(top_task, ctrl_instance_name, v2s, s2v)
+
+  # extract the pblock definition tcl for each region
+  region_to_pblock_tcl = _get_pblock_tcl(topology)
+
+  # determine partitioning of each FIFO, assign each pipeline register to a region
+  fifo_pipeline_level, fifo_pipeline_reg_to_region = get_fifo_pipelines(
+    top_task, v2s, topology)
+  inst_to_region = {**v2s, **fifo_pipeline_reg_to_region}
+
+  return inst_to_region, fifo_pipeline_level, region_to_pblock_tcl
+
+
+def get_post_synth_area(
+    work_dir,
+    cpp_dir,
+    rtl_dir,
+    directive,
+    top_task,
+    task_getter: Callable[[str], Task],
+):
+  def worker(module_name: str, idx: int) -> report.HierarchicalUtilization:
+    _logger.debug('synthesizing %s', module_name)
+    rpt_path = f'{work_dir}/report/{module_name}.hier.util.rpt'
```

Comment:
I think we should assemble `rpt_path` from `tapa.core` and pass that as an argument (instead of passing `work_dir` and assemble the path here). If we ever want to change the directory hierarchy, it'll be easier to have everything in a single module.